### ISSUE TITLE
idea to use homogeneous datamodel

### ISSUE THREAD

**Issue by pointhi** [searx/searx#449]
In my optinion, the current datamodel hase some problems for further applications/better search results:
- we are not able to merge infoboxes
- we are not able to merge infoboxes with normal results
- we cannot represent logical connections (wiki article and official website)
- we cannot merge results in a clean way (keys can be lost, when maps are merged, we cannot merge/display results which are meaning the same thing, but splitted in parts (for example osm streets))

My Idea would be using a datamodel designed to create structured datamodels for the web, more specific I think the schema described on https://schema.org/ should be sufficient for 99% of all possible uses. If done right, searx could also handle databases in much better way (as currently done with wikidata).

I started brainstorming, but there has to be done much more to get a suitable datamodel for searx (what keys are supported,... I started with RDF, but schema.org is much better because it provide a full set of usable keys instead of defining our own or searching for something useful)

https://github.com/pointhi/searx/wiki/idea-for-using-structured-data-to-represent-search-results

When using that model, we could remove the  `template` parameter and replacing it with `'@type'`, and the template would instead aggregate the results in the best way. (for example multible images and some text, and a little map). If done correct this would't require much more template code.

That's only a rough idea, but when done correct we could probably merging/display all kind of data from very different sources, and display them in a very logical/connected way. (As long as searx understand the keys). The disadvantage could be little increased computation time, but when done correctly this shouldn't be a problem.

Feel free to improve my current idea. There are many unclear things left.

**Comment by Cqoicebordel** [comment]
I have not looked at this idea recently, but I was partway through implementing a dynamic template a while back.
And the issue is that it's very difficult. Not impossible, but demanding a lot of calculations and recursive algorithms to go deep in the trees of possible layouts.

I'm not saying that what you are proposing is wrong or unhelpful. On the contrary, I think it would help us do a better work on the backend.
But I also think that it will be the easy part, and that building the corresponding layout won't be that easy.

But we should do it :)

**Comment by pointhi** [comment]
yep, I think this is something which could be started at the earliest after refactoring. I also think we have to design this idea very precisely before starting to code, to have a powerful datamodel which could on the other side also be handled by searx without too much calculations. (Using only a subset of schema.org, defining trees which are valid and easy to handle (for example: how to represent multible wikipedia results, wikidata, the official website and a geoposition in an easy parsable way, where the template can figure out what is important and what is not so important)).

**Comment by Cqoicebordel** [comment]
We had started something similar a while back : https://github.com/Cqoicebordel/searx/wiki/Format-of-answer
And for the full data on building an adaptable theme : https://github.com/Cqoicebordel/searx/wiki/Meta-Theme

**Comment by pointhi** [comment]
yes, but I think my idea goes further, because it allowes merging of results with mutlible urls -> harder to display and parse.

I think after core refactoring is done I'm starting to improve the js and theme part of oscar, to allow more different result types and trying to build a unified result template (for the current data model, but expandable for the future).

Probably I can create some mockups to show possible arrangements of complex results.

**Comment by Cqoicebordel** [comment]
Yes yes, I'm agreeing with you. I think indeed building upon schema.org is a good idea. I just wanted to point out that _some_ of the work already done can be reused (notably the bin packing).

**Comment by asciimoo** [comment]
+1, semantic annotation of the search data would be a huge improvement. I don't have too much xp in these questions, but schema.org looks a good way. What are the next steps?

**Comment by pointhi** [comment]
ok, i have rewritten the wiki page and improved the schema and the data model a bit.

@asciimoo: For the next steps, I think rewrite the engines to the new data model would be the first and easiest task, because it's renaming the keys and adding sub-nodes.
The template and merging code would also need some rework to work with the new keys and nodes, but this would be not so hard, and could already improve searx (for example, we would now be able to store height and width of every image and of it's thumbnail)

With the biggest task (handling of schema, data-model, merging) I would start programming soon, to test my current ideas described [on my wiki page](https://github.com/pointhi/searx/wiki/idea-for-using-structured-data-to-represent-search-results). I already have some ideas how to work effeciently with such data, and how to work with the fuzziness of some datasets (For example, a developer in Wikidata could be either a person or a organisation).

---

One Idea I had would be adding the possibility of listeners, which are triggered when specific datasets are appear. This would allow us to search in general databases like wikidata, and extending the dataset dynamicaly when data is found.
In other words, searx would be able to improve semantic datasets on the fly by looking into additional data sources.

One (very special) example would be a search for NOAA-15 (a weather satellite). When searx notices it found a satellite, it can improve the result by querying additional databases for satellites like the new database of satnogs: https://db.satnogs.org. Also searching in local databases could be triggered by such an event. Theroretical, searx would be capabel of working with thousands of different data sources, without calling all of them for every query.
